presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,225 @@
+            FilterNode filterNode = createFilterNodeForIntersect(aggregationNode, aggregationsBuilder.build());
+
+            return new ProjectNode(idAllocator.getNextId(), filterNode, outputProjections.build());
+        }
+
+        private List<Symbol> generateOutputSymbolsFrom(List<Symbol> sourceSymbols)
+        {
+            ImmutableList.Builder<Symbol> outputSymbolBuilder = ImmutableList.builder();
+            for (Symbol symbol : sourceSymbols) {
+                outputSymbolBuilder.add(allocateNewSymbolFrom(symbol));
+            }
+            return outputSymbolBuilder.build();
+        }
+
+        private Symbol allocateNewSymbolFrom(Symbol symbol)
+        {
+            return symbolAllocator.newSymbol(symbol.getName(), symbolAllocator.getTypes().get(symbol));
+        }
+
+        private AggregationNode createAggregationNodeForIntersect(IntersectNode node, List<Symbol> sourceOutputSymbols, UnionNode sourceNode, ImmutableMap.Builder<Symbol, FunctionCall> aggregationsBuilder, ImmutableMap.Builder<Symbol, Expression> outputProjections)
+        {
+            ImmutableMap.Builder<Symbol, Signature> functionsBuilder = ImmutableMap.builder();
+            ImmutableList.Builder<Symbol> groupByKeys = ImmutableList.builder();
+
+            List<Symbol> intersectOutputSymbols = node.getOutputSymbols();
+            Signature countSignature = new Signature("count", AGGREGATE, "bigint", "boolean");
+
+            for (int i = intersectOutputSymbols.size(); i < sourceOutputSymbols.size(); i++) {
+                Symbol countSymbol = symbolAllocator.newSymbol("count", BIGINT);
+                aggregationsBuilder.put(countSymbol, new FunctionCall(QualifiedName.of("count"), ImmutableList.<Expression>of(new QualifiedNameReference(sourceOutputSymbols.get(i).toQualifiedName()))));
+                functionsBuilder.put(countSymbol, countSignature);
+            }
+
+            for (Symbol nodeSymbol : intersectOutputSymbols) {
+                int fieldIndex = intersectOutputSymbols.indexOf(nodeSymbol);
+                Symbol outputSymbol = sourceOutputSymbols.get(fieldIndex);
+                groupByKeys.add(outputSymbol);
+                outputProjections.put(nodeSymbol, new QualifiedNameReference(outputSymbol.toQualifiedName()));
+            }
+
+            return new AggregationNode(idAllocator.getNextId(),
+                    sourceNode,
+                    groupByKeys.build(),
+                    aggregationsBuilder.build(),
+                    functionsBuilder.build(),
+                    ImmutableMap.<Symbol, Symbol>of(),
+                    ImmutableList.of(groupByKeys.build()),
+                    Step.SINGLE,
+                    Optional.empty(),
+                    1.0,
+                    Optional.empty());
+        }
+
+        private ProjectNode createProjectNode(IntersectNode intersectNode, List<PlanNode> sources, int sourceIndex)
```

Comment:
The structure and signature of this method is a bit complicated. For example, it doesn't really care about all the other entries in `sources` (just the total count), and it doesn't care about `intersectNode` (just the projections).

I'd rewrite it like this:

```java
private ProjectNode createProjectNode(PlanNode source, Map<Symbol, QualifiedNameReference> projections, int markerIndex, int totalMarkers)
{
    ImmutableMap.Builder<Symbol, Expression> assignments = ImmutableMap.builder();
    assignments.putAll(projections);

    // add extra marker fields to the projection
    for (int i = 0; i < totalMarkers; ++i) {
        Symbol output = symbolAllocator.newSymbol(INTERSECT_MARKER, BOOLEAN);
        if (i == markerIndex) {
            assignments.put(output, TRUE_LITERAL);
        }
        else {
            assignments.put(output, new Cast(new NullLiteral(), StandardTypes.BOOLEAN));
        }
    }

    return new ProjectNode(idAllocator.getNextId(), source, assignments.build());
}
```